Fix #4795: Refactor and align SourceRoot.saveAll path resolution with Path.resolve #4795

Joyce-5 opened the pull request:
Fixes https://github.com/javaparser/javaparser/issues/4795.
Refactored and fixed SourceRoot.saveAll to ensure consistent path resolution for both relative and absolute paths.

Changes:

Extracted resolveSavePath(Path newRoot, Path key) as a package visible instance method.

Commit 1: refactor only.

Commit 2: align with java.nio.file.Path#resolve semantics.

Relative keys : resolved under newRoot

Absolute keys : returned unchanged 

Added unit tests for both cases.

Joyce-5 committed:
refactor(utils): extract package-visible instance resolveSavePath and use it in saveAll test(utils): add refactor-only tests documenting current path resolution behaviour #4795

Joyce-5 committed:
fix(SourceRoot): make saveAll(newRoot) always write under newRoot; re-root absolute keys under old root and derive stable path for outside keys using package/primary type or filename; remove Optional parameter usage #4795

jlerbsc commented on javaparser-core/src/main/java/com/github/javaparser/utils/SourceRoot.java:
```diff
@@ -484,16 +486,33 @@ private SourceRoot save(CompilationUnit cu, Path path, Charset encoding) {
+     * @param key     the cached path key (relative or absolute)
+     * @return newRoot.resolve(key) (Path.resolve semantics)
+     */
+    Path resolveSavePath(Path newRoot, Path key) {
```
Please rename the key parameter.

jlerbsc commented on javaparser-core-testing/src/test/java/com/github/javaparser/utils/SourceRootPathResolutionTest.java:
```diff
@@ -11,36 +12,53 @@
-        Path key = Paths.get("pkg").resolve("A.java");
+    @Test
+    void absolute_is_returned_unchanged_pathResolveBehaviour() throws Exception {
+        Path oldRoot = Files.createTempDirectory("jp-old-");
```
I think there is no need to create a temp directory.

jlerbsc commented:
Please take these initial comments into account and squash your commits to make them more readable.

Joyce-5 committed:
fix(utils): align saveAll path computation with java.nio.file.Path#resolve semantics #4795

- rename resolveSavePath to resolvePath and parameter to cachedPath
- simplify logic: relative keys resolved under newRoot; absolute keys returned unchanged (normalized)
- move unit tests into SourceRootTest and simplify setup (no temp directories)
- update comments for clarity

Joyce-5 commented:
I’ve updated the PR based on your feedback.

Renamed resolveSavePath to resolvePath and parameter key to cachedPath

Moved tests into SourceRootTest and simplified them (no temp directories)

Squashed commits into a single clean commit

All tests pass locally, and CI is pending. Please let me know if any further changes are needed.

jlerbsc commented:
Please update your fork with the current version.

Joyce-5 commented:
Thank you very much for your feedback!
To make the new changes based on the current version, I’ve opened a new PR for further updates.
I really appreciate your guidance and time reviewing it.
